src/rule_definition_tools.jl

```diff
@@ -215,14 +219,22 @@ end
 "Declares properly hygenic inputs for propagation expressions"
 _propagator_inputs(n) = [esc(gensym(Symbol(:Δ, i))) for i in 1:n]
 
+"given the variable names, escaped but without types, makes setup expressions for projection operators"
+function _make_projectors(xs)
+    ps = map(x -> gensym(Symbol(:proj_, x.args[1])), xs)
+    setups = map((x,p) -> :($p = ProjectTo($x)), xs, ps)
+    return ps, setups
+end
+
 """
-    propagation_expr(Δs, ∂s, _conj = false)
+    propagation_expr(Δs, ∂s, [_conj = false, proj = identity])
 
-    Returns the expression for the propagation of
-    the input gradient `Δs` though the partials `∂s`.
-    Specify `_conj = true` to conjugate the partials.
+Returns the expression for the propagation of
+the input gradient `Δs` though the partials `∂s`.
+Specify `_conj = true` to conjugate the partials.
+Projector `proj`, usually `ProjectTo(x)`, is applied at the end.
 """
-function propagation_expr(Δs, ∂s, _conj = false)
+function propagation_expr(Δs, ∂s, _conj = false, proj = identity)
```

Comment:
So the reason we have this identity default  is because we are not doing anything at all to it in forwards mode?
Because we having finished thinking that through yet?

Reply:
Yes, haven't thought at all. 

Are there any consumers of `frules` right now? Maybe ForwardDiff2 would be sufficient to see things.

Reply:
Diffractor forwards mode uses frules.

Reply:
I'm not one of the cool people. If you do the equivalent of `gradient(abs2∘/, 1+im, 2)`, do you get 2 complex numbers?

Maybe what I don't see clearly is this: if you start with`x, dx` and push that forwards N steps, maybe you now have `z + dx + dy` I mean like `Dual(z, (dx, dy))`, is the relevant space at this step the one from `z` (maybe the AD does that, ForwardDiff does) or the one from `x`?